A few more bug fixes and some fuzzing stuff

fitzgen opened the pull request:
@cfallin mind taking a look at this one too?

fitzgen committed:
Fix deferring zero-sized allocations to the inner allocator

Forgot a `.inner` in there.

fitzgen committed:
Fix bug in "growing" an allocation smaller than the original allocation

Usually we would be able to reuse the existing allocation, but this is not
possible when the new alignment is not satisfied by the original layout.

fitzgen committed:
Tweak a failing alignment assertion that did not always hold true

It is not necessarily the case that the original layout's alignment is greater
than or equal to the user's layout's alignment. We may have reused an allocation
that happened to be aligned greater than its original layout's requested
alignment such that it satisfied a new user layout's alignment. Therefore, we
instead assert that the pointer is aligned as expected.

fitzgen committed:
Add a bunch more debug assertions to catch potential bugs early

fitzgen committed:
Add a bunch more logging to the fuzz oracle for easier debugging

fitzgen committed:
Use ASan (the default sanitizer) in CI fuzzing

fitzgen committed:
Also build and test the fuzzing utility crate in CI

fitzgen committed:
Fix fuzzing utility crate's CI job; it requires nightly

fitzgen committed:
Cache `cargo-fuzz` install in CI

fitzgen committed:
Fix another alignment debug assertion

fitzgen committed:
Upload fuzzing artifacts in CI for post-mortem debugging

fitzgen committed:
Disable ASan's leak detection in CI

I can't reproduce locally, no matter how many runs I do, and I sort of suspect
this is related to `lazy_static!`-style things in the mutator.

cfallin commented on src/zero_aware_allocator.rs:
```diff
@@ -385,9 +471,16 @@ enum PreGrow {
+    /// less-than-requested alignment, but its actual size is *larger* than the
+    /// new layout's size. Therefore, we need to use the inner allocator to
+    /// shrink instead of grow.
+    DoShrink { actual_old_layout: Layout },
```
I think I believe this can happen, just based on the asserted inequalities, but it would be very helpful to have an example sketched out in this doc-comment (no need for fancy diagrams, just some numbers would do).

The requested alignment *changing* during growth seems to be key -- the if-condition above on line 459 is false if we get `DoShrink`, so `new_layout.size() < node.layout.size()`, so in the compound condition on line 426-427, the second condition (re: alignment) must be false for us to get past that early return.

So basically, we're asking the inner allocator for a new allocation, and it happens to be smaller than the original-with-slop but still bigger than the user's original (because that's what happened to be requested), but the reason for the new allocation is better alignment.

All makes sense, just very subtle -- I think making the "changing allocation" part explicit in the docs is key to avoid the puzzling-out I did here.

fitzgen replied:
> The requested alignment _changing_ during growth seems to be key

Indeed. It is super annoying, as an implementor, that grow/shrink et al can change alignment.

cfallin reviewed (approved):
I think this looks right -- just a comment below asking for a little more doc-commenting!

fitzgen committed:
Pass correct flag for limited-time fuzzing in CI

fitzgen committed:
Add an example to `PreGrow::DoShrink`'s docs

fitzgen commented:
Okay I can repro the leaking on my linux machine actually. Going to investigate some more.

ASan still isn't giving symbolicated stack traces, but what can you do.

fitzgen committed:
Make sure to free the old allocation on successful `grow_zeroed`

When we make a new allocation, rather than growing in place, to satisfy a growth
request, we take ownership of the old allocation and therefore need to free
it. We were correctly doing this for `grow` but not for `grow_zeroed`.

fitzgen committed:
Remove disabling of leak detection from CI fuzzing

fitzgen commented:
Okay I fixed the leak. Dumb.

Care to take a look at the most recent commits again @cfallin?

cfallin reviewed (approved):
Ah, yes, I missed that on earlier review; thank goodness for ASan. LGTM!

cfallin commented:
(Or leak detection in general I suppose.)
